Trigger server

FrankPetrilli opened the pull request:
- We'll deploy it with a on-start generated shared secret, which can be retrieved from the master node.
- It'll be packaged as a systemd .service file so that it runs on startup, etc.
- Packaged as RPM/DEB via FPM

Tests now run successfully with this version.

https://github.com/ipfs/ipfs-cluster/issues/244

FrankPetrilli committed:
https://github.com/ipfs/ipfs-cluster/issues/244

We'll deploy it with a on-start generated shared secret, which can be retrieved from the master node.
It'll be packaged as a systemd .service file so that it runs on startup, etc.
Packaged as RPM/DEB via FPM

Tests now run successfully with this version.

FrankPetrilli commented:
RPM/DEB build artifacts:

```
added QmXrSm9mp3ewBFkUWLbdkkn8f6fwg3SroAbRNj9FumJKbN triggerserver-1.0.0-1.x86_64.rpm
added QmWRWsubcxEpsPwfkSQ62wdrbuWsb3UB3613XaiMCibfYK triggerserver_1.0.0_amd64.deb
```

hsanjuan commented on ipfs-cluster/trigger-server/build.sh:
```diff
@@ -0,0 +1,16 @@
+fi
+go build triggerserver.go
+docker run -v $(pwd):/data --rm fpm /bin/sh -c "cd /data && fpm -s dir -t rpm -n triggerserver -v $VERSION -p /data/ --deb-no-default-config-files triggerserver=/usr/bin/ triggerserver.service=/usr/lib/systemd/system/"
+docker run -v $(pwd):/data --rm fpm /bin/sh -c "cd /data && fpm -s dir -t deb -n triggerserver -v $VERSION -p /data/ --deb-no-default-config-files triggerserver=/usr/bin/ triggerserver.service=/lib/systemd/system/"
```
it is customary to add custom units to /etc/systemd/system/ and leave /lib/systemd for the system.

hsanjuan commented on ipfs-cluster/trigger-server/triggerserver.go:
```diff
@@ -0,0 +1,95 @@
+    "bytes"
+)
+
+const CONFIG_PATH = "/etc/trigger-server"
```
`golint` this file. It's not Go-ish to use all caps in variables

hsanjuan commented on ipfs-cluster/trigger-server/triggerserver.go:
```diff
@@ -0,0 +1,95 @@
+const PORT = 8082
+var RUNNER_PATH = os.Getenv("GOPATH") + "src/github.com/ipfs/kubernetes-ipfs/ipfs-cluster/"
+// Runner takes two args
+const RUNNER_NUM_NODES = "5"
```
Trigger server should be a kubernetes-ipfs wide application that can be used by other projects than ipfs-cluster. Should have a configuration file where the parameters used by each project can be configured

hsanjuan commented on ipfs-cluster/trigger-server/triggerserver.go:
```diff
@@ -0,0 +1,95 @@
+    if req.Method == "POST" {
+	if req.Header.Get("X-Auth") != secret {
+		log.Println("User failed authentication")
+		fmt.Fprintf(rw, "Invalid authentication\n")
```
Set the right http code, 401 indicates unauthorized

hsanjuan replied:
and return so you don't have the whole function nested in an if

hsanjuan commented on ipfs-cluster/trigger-server/triggerserver.go:
```diff
@@ -0,0 +1,95 @@
+		}
+	}
+    } else {
+	// For non-POST requests, tell the user what we are.
```
Check abort conditions at the beginning and return, so the whole body of the funtion is not nested

hsanjuan commented on ipfs-cluster/trigger-server/triggerserver.go:
```diff
@@ -0,0 +1,95 @@
+		// This command takes a while, so we'll hang out here while it runs
+		cmd := exec.Command("bash", "runner.sh", RUNNER_NUM_NODES, RUNNER_NUM_PINS)
+		cmd.Dir = RUNNER_PATH
+		out, err := cmd.CombinedOutput()
```
find a way to stream the output of the command as it happens. Seems https://golang.org/pkg/os/exec/#Cmd.StdinPipe might help. May want to run with a command context so you can cancel the command if you cannot write anymore.

Also, should make sure the deployment is cleaned up from the cluster after the tests have run so they can run again without finding anything dirty.

hsanjuan commented on ipfs-cluster/trigger-server/triggerserver.go:
```diff
@@ -0,0 +1,95 @@
+}
+
+func main() {
+    secret_bytes, readErr := ioutil.ReadFile(CONFIG_PATH)
```
since the config file needs several sections, probably json is a good format.

hsanjuan commented on ipfs-cluster/trigger-server/triggerserver.go:
```diff
@@ -0,0 +1,95 @@
+    secret_bytes, readErr := ioutil.ReadFile(CONFIG_PATH)
+    if readErr != nil {
+        log.Printf("Missing secret configuration in %s. We'll generate one for you...\n", CONFIG_PATH)
+	rand.Seed(time.Now().UnixNano())
```
customary to initialize seeds in init(). Might as well use a proper random source from `crypto` package. Just need to read enough bytes into a buffer and SHA it for example.

hsanjuan commented on ipfs-cluster/trigger-server/triggerserver.go:
```diff
@@ -0,0 +1,95 @@
+	}
+    }
+    secret = string(secret_bytes)
+    log.Printf("Starting with secret as %s on port %d\n", secret, PORT)
```
better not log secrets

hsanjuan commented on ipfs-cluster/trigger-server/triggerserver.go:
```diff
@@ -0,0 +1,95 @@
+	https://github.com/ipfs/kubernetes-ipfs/
+	`)
+    })
+    log.Println(http.ListenAndServe(":" + strconv.FormatInt(PORT, 10), nil))
```
`fmt.Sprintf("%d", PORT)`. Saves one import.

hsanjuan reviewed (changes requested):
In this form the server still suffers the problem that it's going to run whatever ipfs/ipfs-cluster docker container there is.

Instead, it should run the tests with the code that it's being currently being tested (in a PR).

As I pointed out somewhere else, that means that a container needs to be built, published, and the deployment file needs to be pointed to it.

FrankPetrilli committed:
Working toward #42

FrankPetrilli committed:
Resolve #46

FrankPetrilli committed:
Resolve #41

FrankPetrilli committed:
Fix #42, Fix #43, Fix #44, Fix #47

Major changes:

- Output is now streamed from the daemon rather than being dumped at the end.
- At the end of the run, data is dumped to IPFS and you are provided an IFPS link to the output.
- `golint` now passes sucessfully
- The server now responds with 401 to unauthorized users
- User can now specify what they want to run
- Config is now in JSON format
- Config is generated as JSON on first load

FrankPetrilli committed:
Merge branch 'master' into frankpetrilli/trigger_improvement

FrankPetrilli commented:
@hsanjuan, will you check my recent changes?

hsanjuan commented on ipfs-cluster/trigger-server/README.md:
```diff
@@ -1,3 +1,9 @@
+# What is trigger server
+
+The trigger server's purpose is defined in ipfs/ipfs-cluster/issues/244. The goal is to create an endpoint
```
That issue is very long, and deals with other things. In any case, the user should not have to look up an issue in a different repository to find out what the project does, when reading's the project readme.

hsanjuan commented on ipfs-cluster/trigger-server/README.md:
```diff
@@ -1,3 +1,9 @@
+The trigger server's purpose is defined in ipfs/ipfs-cluster/issues/244. The goal is to create an endpoint
+which can be triggered remotely and automated to run the tests, obtaining their results.
+
+
```
Explain that it runs in the kubernetes master? Explain the authentication mechanism and how to set it up? Explain how to put it there after you have the packages? explain that it includes a systemd unit file? how to obtain credentials or where to reach out if needed...

hsanjuan commented on ipfs-cluster/trigger-server/README.md:
```diff
@@ -1,3 +1,9 @@
+
 To build RPM/DEB
 ---
 
```
As an additional note, we don't need new issues for every commit addressing issues in a PR. Commit messages should be more descriptive than this, but it's fine too if you just write `WIP`, and at the end of the review process squash it all together in a single commit, but in any case an additional issue is not needed.

hsanjuan commented on ipfs-cluster/trigger-server/triggerserver.go:
```diff
@@ -8,88 +8,133 @@ import (
+    // This command takes a while, so we'll hang out here while it runs
+    cmd := exec.Command("bash", "-c", fmt.Sprintf("%s%s %s", config.RunnerPath, config.Runner, config.RunnerArgs))
+    cmd.Dir = config.RunnerPath
+    pipeReader, pipeWriter := io.Pipe()
```
Perhaps implement a custom io.Writer. The `Write` method in it just writes the bytes to both the response writer and your `out` buffer. That way you don't need a go-routine on the side reading from one reader and writing somewhere else.

hsanjuan commented on ipfs-cluster/trigger-server/triggerserver.go:
```diff
@@ -8,88 +8,133 @@ import (
+
+func init() {
+    rand.Seed(time.Now().UnixNano())
+    configFile, readErr := ioutil.ReadFile(configPath)
```
configuration handling to ->`main()`, but do create functions for it.. `DefaultConfig()`, `ReadConfig()`, `WriteConfig()`.. they can live in `config.go`

hsanjuan commented on ipfs-cluster/trigger-server/triggerserver.go:
```diff
@@ -8,88 +8,133 @@ import (
+	cwd, _ := os.Getwd()
+	config.Secret = randLetters(30)
+	config.Port = 8082
+	config.RunnerPath = cwd
```
probably need a way to specify several runners for several projects

hsanjuan commented on ipfs-cluster/trigger-server/triggerserver.go:
```diff
@@ -8,88 +8,133 @@ import (
-        fmt.Fprintf(rw, `Kubernetes-IPFS Trigger Server
-	https://github.com/ipfs/kubernetes-ipfs/
-	`)
+	    json.Unmarshal(configFile, &config)
```
error check

hsanjuan replied:
and I think golint would ask you to return from the `if` and not have an `else` clause.
